packages/ceramic-common/src/doctype.ts

```diff
@@ -0,0 +1,228 @@
+
+    /**
+     * Deserializes doctype state from over the network transfer
+     * @param state - Doctype state
+     */
+    static deserializeState(state: any): DocState {
+        state.log = state.log.map((cidStr: string): CID => new CID(cidStr))
+        if (state.anchorProof) {
+            state.anchorProof.txHash = new CID(state.anchorProof.txHash);
+            state.anchorProof.root = new CID(state.anchorProof.root);
+        }
+
+        let showScheduledFor = true;
+        if (state.anchorStatus) {
+            state.anchorStatus = AnchorStatus[state.anchorStatus];
+            showScheduledFor = state.anchorStatus !== AnchorStatus.FAILED && state.anchorStatus !== AnchorStatus.ANCHORED
+        }
+        if (state.anchorScheduledFor) {
+            if (showScheduledFor) {
+                state.anchorScheduledFor = Date.parse(state.anchorScheduledFor); // ISO format of the UTC time
+            } else {
+                state.anchorScheduledFor = null;
+            }
+        }
+        return state
+    }
+}
+
+/**
+ * Doctype init options
+ */
+export interface InitOpts {
+    owners?: Array<string>;
+    onlyGenesis?: boolean;
+    onlyApply?: boolean;
+    skipWait?: boolean;
+    isUnique?: boolean;
+}
+
+/**
+ * Describes common doctype attributes
+ */
+export class Doctype extends EventEmitter {
+    constructor(private _state: DocState) {
+        super()
+    }
+
+    get id(): string {
+        return DoctypeUtils.createDocId(this.state.log[0])
+    }
+
+    get doctype(): string {
+        return this._state.doctype
+    }
+
+    get content(): any {
+        return cloneDeep(this.state.content)
+    }
+
+    get owners(): Array<string> {
+        return cloneDeep(this.state.owners)
+    }
+
+    get state(): DocState {
+        return cloneDeep(this._state)
+    }
+
+    set state(state: DocState) {
+        this._state = state
+    }
+
+    get head(): CID {
+        return this.state.log[this.state.log.length - 1]
+    }
+
+}
+
+/**
+ * Doctype decorator
+ * @constructor
+ */
+export function DoctypeStatic<T>() {
+    return <U extends T>(constructor: U): any => { constructor };
+}
+
+/**
+ * Doctype static signatures
+ */
+export interface DoctypeConstructor<T extends Doctype> {
+    /**
+     * Constructor signature
+     * @param state - Doctype state
+     */
+    new (state: DocState): T;
+
+    /**
+     * Makes genesis record
+     * @param params - Create parameters
+     * @param context - Ceramic context
+     * @param opts - Initialization options
+     */
+    makeGenesis(params: Record<string, any>, context?: Context, opts?: InitOpts): Promise<Record<string, any>>;
+
+    /**
+     * Makes a change on an existing document
+     * @param doctype - Doctype instance
+     * @param params - Change parameteres
+     * @param context - Ceramic context
+     * @param opts - Initialization options
+     */
+    change(doctype: Doctype, params: Record<string, any>, context: Context, opts?: InitOpts): Promise<T>;
```

Comment:
Not sure I understand why this needs to be a static function. Lets talk about it tomorrow!

Reply:
the only reason for `change` to be static is to enforce the result be of type `T extends Doctype`

Reply:
moved it to `Doctype` instance level